central/processlisteningonport/datastore/pruning.go

```diff
@@ -0,0 +1,32 @@
+package datastore
+
+const (
+	// Explain Analyze indicated that 2 statements for PLOP is faster than one.
+	deleteOrphanedPLOPDeploymentsAndPI = `DELETE FROM listening_endpoints WHERE processindicatorid in (SELECT id from process_indicators pi WHERE NOT EXISTS
+                (SELECT 1 FROM deployments WHERE pi.deploymentid = deployments.Id) AND
+                (signal_time < now() at time zone 'utc' - INTERVAL '%d MINUTES' OR signal_time is NULL))`
+
+	deleteOrphanedPLOPPods = `DELETE FROM listening_endpoints WHERE processindicatorid in (SELECT id from process_indicators pi WHERE NOT EXISTS
+                (SELECT 1 FROM pods WHERE pi.poduid = pods.Id) AND
+                (signal_time < now() at time zone 'utc' - INTERVAL '%d MINUTES' OR signal_time is NULL))`
+
+	// Unfortunately if a listening endpoint is marked as being open there is no indication of how old it is.
+	// This leads to a possible race condition where a listening endpoint reaches the database before the deployment,
+	// and the pruning job happens to run before the deployment information arrives in the database.
+	// This should be rare, so this should be acceptable. This could be improved by adding a timestamp to the listening endpoints table
+	deleteOrphanedPLOPDeployments = `DELETE FROM listening_endpoints WHERE NOT EXISTS
+                (SELECT 1 FROM deployments WHERE listening_endpoints.deploymentid = deployments.Id)`
+
+	// Unfortunately if a listening endpoint is marked as being open there is no indication of how old it is.
+	// This leads to a possible race condition where a listening endpoint reaches the database before the pod,
+	// and the pruning job happens to run before the pod information arrives in the database.
+	// This should be rare, so this should be acceptable. This could be improved by adding a timestamp to the listening endpoints table
+	deleteOrphanedPLOPPodsWithPodUID = `DELETE FROM listening_endpoints WHERE poduid IS NOT NULL AND NOT EXISTS
+                (SELECT 1 FROM pods WHERE listening_endpoints.poduid = pods.Id)`
+
+	// Delete orphaned PLOPs
+	pruneOrphanedPLOPs = `DELETE FROM listening_endpoints WHERE closetimestamp < now() at time zone 'utc' - INTERVAL '%d MINUTES'`
+
+	getPotentiallyOrphanedPLOPs = `SELECT plop.serialized FROM listening_endpoints plop LEFT OUTER JOIN process_indicators proc
```

Comment:
What is this one doing?  the word "Potentially" makes me curious.  Probably need a comment here at a minimum.

Reply:
I have added a comment to explain this.

Reply:
So I was curious about this
```
central_active=# explain analyze SELECT plop.id as plop_id FROM listening_endpoints plop where NOT EXISTS (select 1 FROM process_indicators proc where plop.processindicatorid = proc.id);
                                                                              QUERY PLAN                                                                               
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Gather  (cost=1000.42..43193.94 rows=1 width=16) (actual time=0.351..67.880 rows=23194 loops=1)
   Workers Planned: 2
   Workers Launched: 2
   ->  Nested Loop Anti Join  (cost=0.42..42193.84 rows=1 width=16) (actual time=0.082..48.734 rows=7731 loops=3)
         ->  Parallel Seq Scan on listening_endpoints plop  (cost=0.00..3459.80 rows=9980 width=32) (actual time=0.018..5.306 rows=8153 loops=3)
         ->  Index Only Scan using process_indicators_pkey on process_indicators proc  (cost=0.42..3.87 rows=1 width=16) (actual time=0.005..0.005 rows=0 loops=24458)
               Index Cond: (id = plop.processindicatorid)
               Heap Fetches: 1098
 Planning Time: 0.260 ms
 Execution Time: 68.888 ms
(10 rows)


central_active=# explain analyze SELECT plop.id as plop_id FROM listening_endpoints plop LEFT OUTER JOIN process_indicators proc ON plop.processindicatorid = proc.id WHERE proc.id IS NULL;
                                                                              QUERY PLAN                                                                               
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Gather  (cost=1000.42..86381.59 rows=1 width=16) (actual time=0.341..84.388 rows=23345 loops=1)
   Workers Planned: 1
   Workers Launched: 1
   ->  Nested Loop Anti Join  (cost=0.42..85381.49 rows=1 width=16) (actual time=0.070..65.512 rows=11672 loops=2)
         ->  Parallel Seq Scan on listening_endpoints plop  (cost=0.00..2201.46 rows=20146 width=32) (actual time=0.017..4.048 rows=12308 loops=2)
         ->  Index Only Scan using process_indicators_pkey on process_indicators proc  (cost=0.42..4.12 rows=1 width=16) (actual time=0.005..0.005 rows=0 loops=24617)
               Index Cond: (id = plop.processindicatorid)
               Heap Fetches: 1141
 Planning Time: 0.454 ms
 Execution Time: 85.443 ms
(10 rows)
```

These query plans are super similar, I wonder if @erthalion has any thoughts

Reply:
I do think the first one is more on par with the rest of the queries in the codebase so as long as there isn't an issue with it, I would prefer the consistency. Plus it seems to be faster anecdotally

Reply:
I made the query like the top one, except it returns the serialized column
```
        getPotentiallyOrphanedPLOPs = `SELECT plop.serialized FROM listening_endpoints plop where NOT EXISTS
                        (select 1 FROM process_indicators proc where plop.processindicatorid = proc.id)`

```

Reply:
> I do think the first one is more on par with the rest of the queries in the codebase so as long as there isn't an issue with it, I would prefer the consistency. Plus it seems to be faster anecdotally

I don't see any issues with the proposed query. The fact that it's slightly faster probably comes from some volatility around statistics: in the `not exists` the `Parallel Seq Scan` get quite good estimation on the number of rows expected, where in the `left join` it's two times off.